src/CoreWCF.Primitives/src/CoreWCF/Security/NegotiationTokenAuthenticator.cs

```diff
@@ -0,0 +1,1056 @@
+                    }
+                }
+            }
+            finally
+            {
+                if (isNegotiationFailure)
+                {
+                    AddNegotiationChannelForIdleTracking();
+                }
+                else if (negotiationState != null && negotiationState.IsNegotiationCompleted)
+                {
+                    RemoveNegotiationChannelFromIdleTracking();
+                }
+            }
+        }
+
+        // negotiation failure methods
+        Message HandleNegotiationException(Message request, Exception e)
+        {
+
+            //SecurityTraceRecordHelper.TraceServiceSecurityNegotiationFailure<T>(
+            //                        EventTraceActivityHelper.TryExtractActivity(request),
+            //                        this,
+            //                        e);
+            return CreateFault(request, e);
+        }
+
+        Message CreateFault(Message request, Exception e)
+        {
+            MessageVersion version = request.Version;
+            FaultCode subCode;
+            FaultReason reason;
+            bool isSenderFault;
+            if (e is SecurityTokenValidationException || e is System.ComponentModel.Win32Exception)
+            {
+                subCode = new FaultCode(TrustApr2004Strings.FailedAuthenticationFaultCode, TrustFeb2005Strings.Namespace);
+                reason = new FaultReason(SR.Format(SR.FailedAuthenticationTrustFaultCode), CultureInfo.CurrentCulture);
+                isSenderFault = true;
+            }
+            else if (e is QuotaExceededException)
+            {
+                // send a receiver fault so that the sender can retry
+                subCode = new FaultCode(DotNetSecurityStrings.SecurityServerTooBusyFault, DotNetSecurityStrings.Namespace);
+                reason = new FaultReason(SR.Format(SR.NegotiationQuotasExceededFaultReason), CultureInfo.CurrentCulture);
+                isSenderFault = false;
+            }
+            else
+            {
+                subCode = new FaultCode(TrustApr2004Strings.InvalidRequestFaultCode, TrustFeb2005Strings.Namespace);
+                reason = new FaultReason(SR.Format(SR.InvalidRequestTrustFaultCode), CultureInfo.CurrentCulture);
+                isSenderFault = true;
+            }
+            FaultCode faultCode;
+            if (isSenderFault)
+            {
+                faultCode = FaultCode.CreateSenderFaultCode(subCode);
+            }
+            else
+            {
+                faultCode = FaultCode.CreateReceiverFaultCode(subCode);
+            }
+            MessageFault fault = MessageFault.CreateFault(faultCode, reason);
+            Message faultReply = Message.CreateMessage(version, fault, version.Addressing.DefaultFaultAction);
+            faultReply.Headers.RelatesTo = request.Headers.MessageId;
+
+            return faultReply;
+        }
+
+        class NegotiationHost //: ServiceHostBase
+        {
```

Comment:
It might be worth making this into an IServiceDispatcher at some future time. I haven't looked into how this is wired up enough to say if that's feasible, but it feels like that's how this should be implemented.

Reply:
The way it's wired up is each Authenticator is independent and has their own service dispatcher, finally tied to Demuxer based on the the filter.. session is set up after sct..  above authenticator is before sct.... don't know if make sense.. we can discuss..